File: apps/desktop/src/services/commands/progress.ts
import { invoke } from "@/services/tauri";

export interface DailyProgressPointDto {
  date: string;
  cardsReviewed: number;
  xpEarned: number;
  goalMet: boolean;
}

export interface GamificationSummaryDto {
  userId: number;
  totalXp: number;
  level: number;
  currentLevelXp: number;
  nextLevelXp: number;
  xpToNextLevel: number;
  currentStreak: number;
  longestStreak: number;
  todayDate: string;
  dailyGoalCards: number;
  todayCardsReviewed: number;
  todayCardsCorrect: number;
  todayXpEarned: number;
  todayGoalMet: boolean;
  weeklyCardsReviewed: number;
  weeklyXpEarned: number;
  totalSessions: number;
  totalCardsReviewed: number;
  totalCardsCorrect: number;
  accuracy: number;
  masteredWords: number;
  weeklyActivity: DailyProgressPointDto[];
}

export function getGamificationSummary(): Promise<GamificationSummaryDto> {
  return invoke<GamificationSummaryDto>("get_gamification_summary");
}

File: apps/desktop/src-tauri/src/dto/progress.rs
use serde::Serialize;

#[derive(Debug, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DailyProgressPointDto {
    pub date: String,
    pub cards_reviewed: i64,
    pub xp_earned: i64,
    pub goal_met: bool,
}

#[derive(Debug, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct GamificationSummaryDto {
    pub user_id: i64,
    pub total_xp: i64,
    pub level: i64,
    pub current_level_xp: i64,
    pub next_level_xp: i64,
    pub xp_to_next_level: i64,
    pub current_streak: i64,
    pub longest_streak: i64,
    pub today_date: String,
    pub daily_goal_cards: i64,
    pub today_cards_reviewed: i64,
    pub today_cards_correct: i64,
    pub today_xp_earned: i64,
    pub today_goal_met: bool,
    pub weekly_cards_reviewed: i64,
    pub weekly_xp_earned: i64,
    pub total_sessions: i64,
    pub total_cards_reviewed: i64,
    pub total_cards_correct: i64,
    pub accuracy: i64,
    pub mastered_words: i64,
    pub weekly_activity: Vec<DailyProgressPointDto>,
}
